batchflow/models/torch/visualization.py

```diff
@@ -287,6 +287,96 @@ def get_gradcam(self, *args, targets=None, feed_dict=None,
         camera = np.uint8(camera * 255)
         return camera
 
+    # Visualize signal propagation statistics
+    def get_signal_propagation(self, model=None, input_shape=None, batch_size=3, input_tensor=None):
+        """ Compute signal propagation statistics of all layers in the network.
+        Brock A. et al "`Characterizing signal propagation to close the performance gap in unnormalized ResNets
+        <https://arxiv.org/pdf/2101.08692.pdf>`_"
+
+        Under the hood, forward hooks are registered to capture outputs of all layers,
+        and they are removed after extraction of all the activations.
+
+        Parameters
+        ----------
+        model : nn.Module
+            Model to base visualizations on.
+        input_shape : sequence
+            Shape of the image to generate.
+        batch_size : int
+            Number of images in one batch.
+        input : Tensor
+            Input tensor for signal propagation.
+        """
+        if model is None:
+            try:
+                model = self.model
+            except:
+                raise ValueError("You should provide TorchModel or nn.Module.")
+
+        if input_tensor is None:
+            if input_shape is None:
+                raise ValueError("You should provide `input_tensor` or `input_shape` argument.")
+            if len(input_shape) != 3:
+                raise ValueError("Parameter `input_shape` must be a sequence of 3 elements (C, H, W).")
+            input_shape = (batch_size, *input_shape)
+            input_tensor = torch.randn(input_shape)
+
+        statistics = {
+            'Average Channel Squared Mean': [],
+            'Average Channel Variance': [],
+            'Modules instances': []
+        }
+        extractors = []
+        signals = []
+
+        try:
+            for module in model.modules():
+                submodules_amount = sum(1 for _ in module.modules())
+                if submodules_amount == 1:
+                    module_instance = module.__class__.__name__
+                    if module_instance != 'Identity':
+                        statistics['Modules instances'].append(module_instance)
+                        extractors.append(LayerExtractor(module))
```

Comment:
This is more important than the previous line; also, you've created `extractors` and `signals` in other order. So, move this line up